docs/language/informal/toplevel-inference.md

```diff
@@ -0,0 +1,293 @@
+  differently at different syntactic positions.  It’s ok for an expression to be
+  inferrable at one level but not at another.
+* Obvious types should be inferred
+* Inferred and annotated types should be treated the same
+* Simple intuition for users
+* Efficient to implement
+* Type arguments should be mostly inferred
+
+
+A motivating example for the current strong mode implementation (that also
+serves as a demonstration of its shortcomings) is as follows:
+
+
+```dart
+class A {
+  var x = 3;
+}
+var a = new A();
+var b = a.x;
+```
+
+
+To ensure stability with respect to cycles and different orderings, current
+strong mode does inference in two passes: first on the static variables, and
+secondly on fields (method override based inference is mostly irrelevant for
+this discussion). The implementation ensures that inference is stable, but it
+has several unfortunate consequences, most notably the following.  Inference may
+result in the same variable being implicitly seen at two different types. In the
+example above, if all declarations are in the same library, then b will be
+inferred before x has been inferred, and so b will be inferred to have type
+dynamic.  The same declaration in a separate library will see the type of a.x
+after inference, and hence will be inferred to have type `int`.  This violates
+one of the principles above, and also means that changes to the library
+structure that introduce a cycle may change inference results.
+
+
+## Method inference
+
+
+### Method inference
+
+
+Method inference for a method m behaves as if all method inference for all of
+its supertypes has already been performed.
+
+
+If a method leaves a type off of its signature (parameter type or return type)
+and it does not override or implement anything from a super type, it is an
+error.
+
+
+If a method leaves a type off of its signature (parameter type or return type)
+and it does not override or implement anything from a super type, then the
+omitted types are treated as dynamic.
+
+
+Otherwise, the missing types are filled in with the type from the overridden
+method.  If there are multiple overridden/implemented methods and the type to be
+filled in does not agree, it is an error.  If there is no corresponding
+parameter position in the overridden method to infer from, it is treated as
+dynamic (e.g. overriding a one parameter method with a method that takes a
+second optional parameter).
+
+
+### Setter return types
+
+
+If the return type is left off a setter, the return type is inferred to be
+`void`.
+
+
+## Static variable and field inference
+
+
+A “candidate” for inference is any of the following:
+* A top level variable definition
+* A static class level variable definition
+* An instance variable, getter or setter definition with the type annotation
+  omitted.
+
+
+The *inference dependencies of a candidate* which is inferred via override
+inference are the instance fields, setters, and getters that the candidate
+overrides.
```

Comment:
The issue is that for fields/getters/setters, we allow initializer inference to feed into override inference, so this must be done separately from method override inference.